### PR TITLE
Adapt rotate changes

### PR DESCRIPTION
**Pull request by Jedi18**

### CODE REVIEW

**Comment by hcq9102** [comment]
> I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> 
> Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.

Thanks a lot!  I'd think you will just check the cpos and Dpos in /.../algorithms/rotate.hpp, So I didnt update my newest changes which you had suggested changes, and some errors there, this leads to you spend more time on the fixed problems.  Sorry for that, won't happen again.

**Comment by hcq9102** [comment]
> I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> 
> Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.

Got it, That's cool !

**Comment by hcq9102** [comment]
> No, this denotes the return type, it has nothing to do with f1 or f2. If you remember, we write a lambda as -> return_type {} . So we are just telling the lambda that we will return a future containing util::in_out_result<FwdIter, Sent>.

make snese!

**Comment by hcq9102** [comment]
> Yes the parameters could be anything, we can change it to Sent.

ok, learned. thx!

**Comment by hcq9102** [comment]
> rotate

So does similar changes in /test/rotate_copy_range.cpp file.

**Comment by hcq9102** [comment]
> I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> 
> Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.

So does similar changes in /test/rotate_copy_range.cpp file.

**Comment by hcq9102** [comment]
> I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> 
> Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.

So do similar changes in /test/rotate_copy_range.cpp file.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 1]
```diff
@@ -76,18 +246,18 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
-            template <typename ExPolicy, typename FwdIter>
+            template <typename ExPolicy, typename FwdIter, typename Sent>
             static typename util::detail::algorithm_result<ExPolicy,
                 IterPair>::type
-            parallel(ExPolicy&& policy, FwdIter first, FwdIter new_first,
-                FwdIter last)
+            parallel(
+                ExPolicy&& policy, FwdIter first, FwdIter new_first, Sent last)
             {
                 return util::detail::algorithm_result<ExPolicy, IterPair>::get(
```
Comment (hcq9102) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:262]: the second parameter of algorithm_result  'IterPair' is fine here?

Comment (Jedi18) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:262]: Yes, if you look at line 242, the iterpair we use here is the one being used. When calling the function in the tag dispatch overloads    
```
parallel::v1::detail::rotate<
hpx::parallel::util::in_out_result<FwdIter, Sent>>()
.call(hpx::execution::seq, first, new_first, last)
```
Here IterPair would refer to hpx::parallel::util::in_out_result<FwdIter, Sent>. So IterPair is just used to refer to the return type.

**Comment by Jedi18** [comment]
> > I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> > Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.
> 
> So do similar changes in /test/rotate_copy_range.cpp file.

I think so yeah, if rotate copy uses reverse too then we'll need to do that. I'll check it later today.

**Comment by hcq9102** [comment]
I already fixed some errors and update the changes on the branch, the check result shows like https://github.com/STEllAR-GROUP/hpx/pull/5459/checks, there is one tests.unit fails, like (https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/8538/workflows/55f8c2b2-9f1a-4fac-b48e-f636c29005b6/jobs/262921), I didn't get it.

I appreciate your help.

**Comment by hcq9102** [comment]
> > > I've removed the exception and badalloc tests and not used the test_iterator and test_container because the parallel implementation of rotate uses reverse (which requires a bidirectional iterator) which is not supported by these test_iterator and test_container.
> > > Once your PR gets merged I'll create a seperate PR to fix this, since rotate should'nt require a bidirectional iterator. We could then add the exception and badalloc tests.
> > 
> > 
> > So do similar changes in /test/rotate_copy_range.cpp file.
> 
> I think so yeah, if rotate copy uses reverse too then we'll need to do that. I'll check it later today.

Thanks a lot! maybe some errors I didn't catch leads to the tests.unit fails.  thanks for your time.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp (hunk 1 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:262]: much clear!

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp [hunk 2]
```diff
@@ -122,26 +365,284 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_range<Rng>::value)>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            subrange_t<typename hpx::traits::range_iterator<Rng>::type,
+                typename hpx::traits::range_sentinel<Rng>::type>>::type
+        tag_fallback_dispatch(hpx::ranges::rotate_t, ExPolicy&& policy,
+            Rng&& rng, typename hpx::traits::range_iterator<Rng>::type middle)
+        {
+            typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
```
Comment (Jedi18) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:535]: This should be typedef std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_bidirectional_iterator<typename hpx::traits::range_iterator<Rng>::type>::value>
                is_seq;

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp [hunk 3]
```diff
@@ -122,26 +365,284 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                Rng>::value&& hpx::traits::is_iterator<OutIter>::value)>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            rotate_copy_result<typename hpx::traits::range_iterator<Rng>::type,
+                OutIter>>::type
+        tag_fallback_dispatch(hpx::ranges::rotate_copy_t, ExPolicy&& policy,
+            Rng&& rng, typename hpx::traits::range_iterator<Rng>::type middle,
+            OutIter dest_first)
+        {
+            return hpx::parallel::v1::detail::rotate_copy<rotate_copy_result<
```
Comment (Jedi18) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:639]: Add typedef std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_forward_iterator<typename hpx::traits::range_iterator<Rng>::type>::value>
                is_seq;

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp [hunk 4]
```diff
@@ -122,26 +365,284 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                typename hpx::traits::range_iterator<Rng>::type, OutIter>>()
+                .call(std::forward<ExPolicy>(policy), hpx::util::begin(rng),
```
Comment (Jedi18) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:641]: use .call2 and pass in is_seq().

Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:641]: > use .call2 and pass in is_seq().

got it.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp (hunk 2 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:535]: my bad. fixed

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp (hunk 3 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:639]: So you mean the following:

`typedef std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_bidirectional_iterator<
                        typename hpx::traits::range_iterator<Rng>::type>::value>
                is_seq;`
the plain text always miss sth.

Comment (Jedi18) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:639]: Yep, sorry should've used the code block while pasting.

Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:639]: > Yep, sorry should've used the code block while pasting.

Thanks a lot, just make sure with you~

**Pull request closed without merging**